### PR TITLE
Cleanup sub fetch phase extension point

### PR DESCRIPTION
**Pull request by javanna**
- move `DocValueFieldsSubFetchPhase` to the same code structure as other sub fetch phases, as it doesn't need to use plugins extensions points
- move `parseElements` method out of `SearchPhase` interface as it's needed only for fetch sub phases
- `SubFetchPhase` doesn't need to provide a map of parsers, one parser is enough
- rename `SearchParseElement` to `SearchExtParser`, change signature of parse method to return what gets parsed (`SearchExtBuilder`) and get only `XContentParser` as an argument. Add `getName()` method to it.
- `SearchSourceBuilder` to parse ext sections straightaway to their own SearchExtBuilder extensions, which are `NamedWriteable` and `ToXContent`.
- `SearchService` to retrieve the parsed builders and set them to the search context for further retrieval
- Remove notion of `SubFetchPhaseContext` (it was only providing `hitsExecutionNeeded` which is not needed and can be removed). What gets parsed is simply an arbitrary object.

### CODE REVIEW

**Comment by javanna** [comment]
@colings86 can you have a look?

File: core/src/main/java/org/elasticsearch/search/SearchExtParser.java [hunk 1]
```diff
@@ -20,12 +20,19 @@
...
+    /**
+     * Returns the name of the element that this parser is able to parse
+     */
+    String getName();
 
-    void parse(XContentParser parser, SearchContext context) throws Exception;
+    /**
+     * Parses the element whose name is returned by {@link #getName()}
+     */
+    Object parse(XContentParser parser) throws Exception;
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/SearchExtParser.java]: Should this not also have access to the `ParseFieldMatcher` so extensions can use the `ParseField` stuff?

Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/SearchExtParser.java]: potentially, but I will remove the parse field matcher very soon, so I would not add it here.

**Comment by colings86** [comment]
I like this change. I wonder though if we should also rename the `ext` in the search request to something less generic and more relevant to adding custom fetch sub phases? Having it as `ext` suggests that you can extend the search request in lots of ways but in actual fact you can only add custom fetch sub phases. I guess this really depends on whether we can think of a better name and whether we see us wanting to add any other extension points to the search request that would use this `ext` object? My feeling is that we wouldn't want to add any other extension points but I don't have fully visibility here. 

@clintongormley @brwe @imotov maybe you have suggestions/thoughts here?

**Comment by brwe** [comment]
`ext` seems rather generic. I cannot think of any other way we would like to extend search requests now and even if we do that would be a new pr anyway since ext is rather specific to fetch phase. I would probably rename to `custom_fetch_element` or something similar.

**Comment by javanna** [comment]
I am fine with renaming. I kinda wish this discussion had happened earlier when `ext` was introduced. That happened for 5.0 though I suspect so no biggie? Were there many more ways to extend a search request at the time?

**Comment by javanna** [comment]
Thinking more about it, maybe ext was a good choice? We cannot change the structure of the json search request that often, so maybe it is ok to have something generic that leaves room for potential additional extensions? That would be the only reason to keep `ext` in the search request json format, although at the moment only sub fetch phases can register parsers for objects under that section.  Or maybe we should decouple the notion SearchExtParser from SubFetchPhase and remove the `SubfetchPhase#parser` method, which is optional anyways and come up with a different way to register an ext parser which will still parse and store the parsed object within the search context. Thoughts?

**Comment by javanna** [comment]
With the last commit I decoupled the parsing of an ext section from sub fetch phases. I think this change clarifies what the extensions points are: 1) sub fetch phase 2) search ext parser that needs to be registered separately

The only thing is that this parsing happens on the data nodes. I am not sure we want to move it to the coordinating node as it would require the parsed object to be a NamedWriteable which makes things more complicated. Let me know what you think.

**Comment by javanna** [comment]
With my last commit I also moved parsing of the ext sections to the coordinating node. It requires plugins to have a builder object that implements `NamedWriteable` and `ToXContent` (actually that extends `SearchExtBuilder`) but I think it makes things cleaner and more inline with how search requests get parsed. This is ready for review. Maybe @nik9000 you want to have a look?

File: core/src/test/java/org/elasticsearch/search/builder/SearchSourceBuilderTests.java [hunk 2]
```diff
@@ -388,13 +389,9 @@ public static SearchSourceBuilder createSearchSourceBuilder() throws IOException
...
             builder.aggregation(AggregationBuilders.avg(randomAsciiOfLengthBetween(5, 20)));
         }
-        if (randomBoolean()) {
-            XContentBuilder xContentBuilder = XContentFactory.jsonBuilder();
-            xContentBuilder.startObject();
-            xContentBuilder.field("term_vectors_fetch", randomAsciiOfLengthBetween(5, 20));
-            xContentBuilder.endObject();
-            builder.ext(xContentBuilder);
-        }
+        //if (randomBoolean()) {
```
Comment (colings86) [inline core/src/test/java/org/elasticsearch/search/builder/SearchSourceBuilderTests.java]: we should uncomment this

**Comment by colings86** [comment]
@javanna I took a quick look and think it looks great, much easier to move over to the coordinating node than I thought it would be. Should we add a note in the breaking changes doc as courtesy? Especially since we have the other java breaking changes in there, and also it might be worth adding a note to the ExtBuilder to say that any state needs to be serialised just as a warning to anyone who extends it?

Still needs a thorough look over form someone else though as I only had time for a quick look

File: core/src/test/java/org/elasticsearch/search/builder/SearchSourceBuilderTests.java (hunk 2 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/builder/SearchSourceBuilderTests.java]: yep leftover

**Comment by nik9000** [comment]
> We cannot change the structure of the json search request that often

I think we can add things to it as we need, right? So long as we don't clash.

File: core/src/main/java/org/elasticsearch/search/SearchExtBuilder.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+import org.elasticsearch.common.io.stream.NamedWriteable;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.plugins.SearchPlugin;
+
+/**
+ * Intermediate serializable representation of a search ext section. To be subclassed by plugins that support
+ * a custom section as part of a search request, which will be provided within the ext element.
+ * Any state needs to be serialized as part of the {@link org.elasticsearch.common.io.stream.Writeable#writeTo(StreamOutput)} method and
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/SearchExtBuilder.java]: I'd just import `Writeable` so the source for this javadoc is easier to read. The written javadoc will be the same but we ofter read the javadoc in source form.

File: core/src/main/java/org/elasticsearch/search/SearchRequestParsers.java [hunk 4]
```diff
@@ -37,28 +37,38 @@
     /**
      * Query parsers that may be used in search requests.
      * @see org.elasticsearch.index.query.QueryParseContext
-     * @see org.elasticsearch.search.builder.SearchSourceBuilder#fromXContent(QueryParseContext, AggregatorParsers, Suggesters)
+     * @see org.elasticsearch.search.builder.SearchSourceBuilder#fromXContent(QueryParseContext, AggregatorParsers,
+     * Suggesters, SearchExtRegistry)
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/SearchRequestParsers.java]: Can you add another tab here?

File: core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 5]
```diff
@@ -203,18 +203,10 @@ public SearchSourceBuilder(StreamInput in) throws IOException {
         postQueryBuilder = in.readOptionalNamedWriteable(QueryBuilder.class);
         queryBuilder = in.readOptionalNamedWriteable(QueryBuilder.class);
         if (in.readBoolean()) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: Can empty list be a valid value here so we can avoid the boolean? Not needed as part of this PR but it'd be nice.

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java [hunk 6]
```diff
@@ -872,6 +872,16 @@ public void writeList(List<? extends Writeable> list) throws IOException {
     }
 
     /**
+     * Writes a list of strings
+     */
+    public void writeStringList(List<String> list) throws IOException {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: Maybe this should just be `public void writeList(Writer<?> list) throws IOException {`?

If we did that we should more `Writer` out of `Writeable` and put it in `StreamOutput` I guess.... And we'd want a twin method in `StreamInput` and to move `Reader` there too I guess....

Maybe that should just wait until another PR.

File: core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 7]
```diff
@@ -1034,8 +1021,23 @@ public void parseXContent(QueryParseContext context, AggregatorParsers aggParser
...
+                    extBuilders = new ArrayList<>();
+                    String extSectionName = null;
+                    while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                        if (token == XContentParser.Token.FIELD_NAME) {
+                            extSectionName = parser.currentName();
+                        } else {
+                            SearchExtParser searchExtParser = searchExtRegistry.lookup(extSectionName,
+                                    context.getParseFieldMatcher(), parser.getTokenLocation());
+                            SearchExtBuilder searchExtBuilder = searchExtParser.fromXContent(parser);
+                            if (searchExtBuilder.getWriteableName().equals(extSectionName) == false) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: Do you think we could get away with making this an assertion instead? A plugin developer would hit it during testing.

File: core/src/test/java/org/elasticsearch/search/fetch/FetchSubPhasePluginIT.java [hunk 8]
```diff
@@ -48,21 +52,27 @@
...
-@ClusterScope(scope = Scope.SUITE, supportsDedicatedMasters = false, numDataNodes = 1)
+@ClusterScope(scope = Scope.SUITE, supportsDedicatedMasters = false, numDataNodes = 2)
 public class FetchSubPhasePluginIT extends ESIntegTestCase {
     @Override
     protected Collection<Class<? extends Plugin>> nodePlugins() {
         return Collections.singletonList(FetchTermVectorsPlugin.class);
     }
 
+    @Override
+    protected Collection<Class<? extends Plugin>> transportClientPlugins() {
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/fetch/FetchSubPhasePluginIT.java]: I really wonder if this should be the default implementation....

**Comment by nik9000** [comment]
I'm still not 100% sold on us needing `ext`. We allow plugins to add new queries, highlighters, aggregations, function_scores, all kinds of stuff. Adding a new top level query term isn't that big a stretch. I guess you could argue that that makes it look like the component is part of Elasticsearch but so does allowing plugins to register queries and stuff. But that is a debate for another time and another place.

LGTM

**Comment by nik9000** [comment]
I left a couple of things that can be follow ups or cleanups. The parsing all makes sense to me and I'm glad we have a plugin to test it.

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java (hunk 6 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: yea let's do it in a separate change, I was annoyed here that I could read a string list in one line but not write it. I made that work.

File: core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java (hunk 7 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: but then you need assertions enabled, I think throwing the exception is fine and makes absolutely sure that things are correct.

File: core/src/test/java/org/elasticsearch/search/fetch/FetchSubPhasePluginIT.java (hunk 8 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/fetch/FetchSubPhasePluginIT.java]: I know I know :)

**Comment by javanna** [comment]
> I'm still not 100% sold on us needing ext. We allow plugins to add new queries, highlighters, aggregations, function_scores, all kinds of stuff. Adding a new top level query term isn't that big a stretch. I guess you could argue that that makes it look like the component is part of Elasticsearch but so does allowing plugins to register queries and stuff. But that is a debate for another time and another place.

I agree, I found out a few days ago that we had to be honest. I think this PR improves things as it decouples it from sub fetch phases as they don't have to come togetherm although at the moment a custom sub fetch phase is the only thing that may need additional config which can be provided via ext.

**Pull request merged**